src/deepsparse/benchmark_model/benchmark.py

```diff
@@ -0,0 +1,138 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import queue
+import threading
+import time
+from typing import Dict, List
+
+import numpy as np
+
+from deepsparse import Engine
+
+
+__all__ = ["model_stream_benchmark"]
+
+
+def iteration(model: Engine, input: List[np.ndarray]):
+    start = time.time()
+    output = model(input)
+    end = time.time()
+    return output, start, end
+
+
+def singlestream_benchmark(
+    model: Engine,
+    input_list: List[np.ndarray],
+    seconds_to_run: float,
+) -> List[float]:
+    batch_times = []
+
+    stream_end_time = time.time() + seconds_to_run
+    while time.time() < stream_end_time:
+        _, start, end = iteration(model, input_list)
+        batch_times.append([start, end])
+
+    return batch_times
+
+
+class EngineExecutorThread(threading.Thread):
+    def __init__(
+        self,
+        model: Engine,
+        input_list: List[np.ndarray],
+        time_queue: queue.Queue,
+        max_time: float,
+    ):
+        super(EngineExecutorThread, self).__init__()
+        self._model = model
+        self._input_list = input_list
+        self._time_queue = time_queue
+        self._max_time = max_time
+
+    def run(self):
+        while time.time() < self._max_time:
+            _, start, end = iteration(self._model, self._input_list)
+            self._time_queue.put([start, end])
+
+
+def multistream_benchmark(
+    model: Engine,
+    input_list: List[np.ndarray],
+    seconds_to_run: float,
+    num_streams: int,
+) -> List[float]:
+    time_queue = queue.Queue()
+    max_time = time.time() + seconds_to_run
+    threads = []
+
+    for thread in range(num_streams):
+        threads.append(EngineExecutorThread(model, input_list, time_queue, max_time))
+
+    for thread in threads:
+        thread.start()
+
+    for thread in threads:
+        thread.join()
+
+    return list(time_queue.queue)
+
+
+def model_stream_benchmark(
+    model: Engine,
+    input_list: List[np.ndarray],
+    scenario: str,
+    seconds_to_run: float,
+    num_streams: int,
+) -> Dict:
+    batch_times = []
+
+    # Run the benchmark scenario and collect batch times
+    if scenario == "singlestream":
+        batch_times = singlestream_benchmark(model, input_list, seconds_to_run)
+    elif scenario == "multistream":
+        batch_times = multistream_benchmark(
+            model, input_list, seconds_to_run, num_streams
+        )
+    else:
+        raise Exception(f"Unknown scenario '{scenario}'")
+
+    # Convert times to milliseconds
+    batch_times_ms = [
+        (batch_time[1] - batch_time[0]) * 1000 for batch_time in batch_times
+    ]
+
+    # Calculate statistics
+    first_start_time = min([b[0] for b in batch_times])
+    last_end_time = max([b[1] for b in batch_times])
+    total_time_executing = last_end_time - first_start_time
```

Comment:
I don't get this bit here.  It seems like this calculation will include all the test overhead in total_time_executing?

Why not use the sum of all the (end-start) times for each batch?

Reply:
We want to know all of the executions that could be performed within a given amount of wallclock time. For multistream if we summed all the times for each batch this would be incorrect. 
This is meant to be a realistic but best case benchmark so I tried to remove overhead from the actual runs, those sections are definitely up to be audited.

Reply:
Ok I guess that's reasonable.  A comment would be good tho.  If this only applicable for multi-stream then maybe we should not penalize ourselves in the singlestream case.  It also would probably be simpler to just record the start and end times around multistream_benchmark/singlestream_benchmark instead?

Reply:
This gives less overhead compared to timing the whole benchmark function call so I'll keep it as is. I left a hefty comment and think we should do measurements of overhead later